app/models/bulk_download.rb

```diff
@@ -25,4 +30,161 @@ def convert_params_to_json
+    command += "--ecr-image #{Shellwords.escape(ecr_image)} "
+    command += "--fargate-cpu #{fargate_cpu} "
+    command += "--fargate-memory #{fargate_memory}"
+    command
+  end
+
+  def s3_tar_writer_command(
+    src_urls,
+    tar_names,
+    dest_url,
+    success_url: self.success_url,
+    error_url: self.error_url,
+    progress_url: self.progress_url
+  )
+    command = "python s3_tar_writer.py "
+    escaped_src_urls = src_urls.map { |url| Shellwords.escape(url) }.join(' ')
+    command += "--src-urls #{escaped_src_urls} "
+    escaped_tar_names = tar_names.map { |url| Shellwords.escape(url) }.join(' ')
+    command += "--tar-names #{escaped_tar_names} "
+    command += "--dest-url #{Shellwords.escape(dest_url)} "
+
+    # The success url is mandatory.
+    if success_url
+      command += "--success-url #{Shellwords.escape(success_url)} "
+    elsif Rails.env == "staging" || Rails.env == "prod"
+      raise BulkDownloadsHelper::SUCCESS_URL_REQUIRED
+    else
+      raise BulkDownloadsHelper::SUCCESS_URL_REQUIRED_LOCAL
+    end
+    if error_url
+      command += "--error-url #{Shellwords.escape(error_url)} "
+    end
+    if progress_url
+      command += "--progress-url #{Shellwords.escape(progress_url)}"
+    end
+    command
+  end
+
+  def download_output_key
+    "downloads/#{id}/#{download_type}.tar.gz"
+  end
+
+  # The s3 url that the tar.gz file will be uploaded to.
+  def download_output_url
+    "s3://#{ENV['SAMPLES_BUCKET_NAME']}/#{download_output_key}"
+  end
+
+  def kickoff_ecs_task(
+    shell_command
+  )
+    aegea_command = aegea_ecs_submit_command(shell_command)
+
+    command_stdout, command_stderr, status = Open3.capture3(aegea_command)
+    if status.exitstatus.zero?
+      output = JSON.parse(command_stdout)
+      # Store the taskArn for debugging purposes.
+      self.ecs_task_arn = output['taskArn']
+      self.status = STATUS_RUNNING
+      save
+    else
+      self.status = STATUS_ERROR
+      self.error_message = BulkDownloadsHelper::KICKOFF_FAILURE
+      save
+      raise command_stderr
+    end
+  end
+
+  def bulk_download_ecs_task_command
+    if download_type == "original_input_file"
+      samples = Sample.where(id: pipeline_runs.map(&:sample_id))
+                      .includes(:input_files)
+
+      download_src_urls = samples.map do |sample|
+        sample.input_files.map { |input_file| "s3://#{ENV['SAMPLES_BUCKET_NAME']}/#{input_file.file_path}" }
+      end.flatten
+
+      # We use the sample name in the input name because it's what is visible to the user.
+      # Also, there might be duplicates in the original file name.
+      download_tar_names = samples.map do |sample|
+        # We assume that the first input file is R1 and the second input file is R2.
+        sample.input_files.map.with_index do |input_file, input_file_index|
+          "#{input_file.sample.name}__original_R#{input_file_index + 1}.#{input_file.file_type}"
```

Comment:
Is there any special reason why not `input_file.name` here instead of trying to guess the name and order?
I guess this is because we can rename samples, but should we rename the "original" files?

I have a suggestion, instead of adding the word `_original` to the file name, what if we store it in a subfolder containing the sample name, like `samples/CURRENT_SAMPLE_NAME/fastqs/ORIGINAL_SAMPLE_NAME_R1.fastq.gz`.

Reply:
That's a great point regarding truncation! Will make a note to add that to the docs...although most users will not be using our S3 upload (that's just for biohub)

Regarding the input-file name, input files can be named the same thing. Multiple samples could have input files named "sample_R1.fastq". Whereas I believe sample names must be unique. Also, it's more intuitive for the user when we name it based on the name of the current sample. They might not remember what the sample was originally called, or what the file was originally named when they uploaded it, so giving them that additional information doesn't seem to add much.

I tried to avoid using subfolders in the tarfile because I worried a more complex directory structure could be more confusing for users to deal with. Tried to keep it simple.

Reply:
sample names are unique per project, so is it possible to select from multiple projects?

Reply:
I personally think folders are more organized than have long names that doesn't reflect the original one. Specially considering what @jshoe just said.
Should we have something like "#{project_name}/#{sample_name}/#{original_file_name}"?
Opinions?
(Note that we may need to escape the project name, since it doesn't have the same constraints as input files have)

Reply:
Let me check with @happyimadesignr. I'm okay with it.

Reply:
I think this decision rests on how users will use our files.

One reason why it might be good for all the files to be in one directory is that it's easier for users to select multiple files to upload to an online tool (like NCBI).

I could see having a bunch of directories being annoying if you want to select multiple files.

Even programmatically, instead of saying "all the files in this directory", you have to do a little extra work to traverse the directories.

Reply:
I can see this is a subjective topic.

The alternative for unique names without folders would probably be something like `<project_name>__<inferred_sample_name>.<file_type>`.

Checked production database and collected some stats:
```
Project.maximum("length(name)")
=> 51
Sample.maximum("length(name)")
=> 127
```

So potentially it could be something like:

`MY_PROJECT_NAME_WITH_51_CHARACTERSXXXXXXXXXXXXXXXX__MY_SAMPLE_NAME_WITH_121_CHARACTERSXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.fastq.gz`

I also collected the average name length, so we can have an informed decision:
```
Project.average("length(name)").to_i
=> 15
Sample.average("length(name)").to_i
=> 28
```
(average may not be the best metric, but just to have an idea)

So it will probably looks like this:
`MY_PROJECT_NAME__MY_SAMPLE_NAME_WITH_28_CHA.fastq.gz`

Reply:
Additionally, it is worth pointing out that we may need to deduplicate project names (doesn't matter if it a folder or part of the file name), because there may have a collision after escaping the file name:

```
PROJECT 1
PROJECT_1
PROJECT*1
```
they all would need to be normalized to something like `PROJECT_1` (we don't want special characters in the file name)

Reply:
Talked to Jenn and Katrina. It seems like scientists are used to long names, and would appreciate having everything in the root directory.

We can do `{sample_name}_{project_name}__original_r1.fastq.gz`. Sample name first so they can sort by file name.

Regarding the fact that we infer which file is "R1" and which one is "R2" based on index in `sample.input_files`, it looks like that's how we handle it when passing files to the pipeline as well. So at least it's consistent.

`pipeline_run_stage.rb`:
```
attribute_dict = {
      fastq1: sample.input_files[0].name,
}
attribute_dict[:fastq2] = sample.input_files[1].name if sample.input_files[1]  
```

Reply:
Just saw your comments above. Thanks for doing those queries!

Reply:
What's the best way to escape the project name? (is there a standard function I can use, or will have to replace a regex with `_`) I think we can do `{sample_name}__{escaped_project_name}__{project_id}__original.fastq.gz`. The project id will ensure uniqueness.

Reply:
I would suggest using something similar to The sample name (https://github.com/chanzuckerberg/idseq-web/blob/8ed4fd810ec1cf66e7c9c1993ad875a5639b1a4c/app/models/input_file.rb#L9   )

Basically anything that is not a letter, number, underscore, minus sign or a dot, should be replaced by an underscore.

We also don’t want to allow the first letter of the name to be a dot or a minus sign because these things have a special meaning in that position (dot is a hidden file, and a minus sign could be interpreted as a parameter by some commands)